lock-api/src/main/java/com/palantir/lock/client/MultiClientTransactionStarter.java

```diff
@@ -0,0 +1,323 @@
+                startTransactionResponses.forEach((namespace, responseList) -> {
+                    multiClientRequestManager.updatePendingStartTransactionsCount(namespace, responseList.size());
+                    namespaceWiseResponseHandler.get(namespace).processResponse(responseList);
+                });
+            }
+        } finally {
+            clearResources(multiClientRequestManager, namespaceWiseResponseHandler);
+        }
+    }
+
+    private static void clearResources(
+            MultiClientRequestManager multiClientRequestManager,
+            Map<Namespace, ResponseHandler> namespaceWiseResponseHandler) {
+        multiClientRequestManager.close();
+        namespaceWiseResponseHandler.forEach((_unused, responseHandler) -> responseHandler.close());
+    }
+
+    private static Map<Namespace, ResponseHandler> getResponseHandlers(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        Map<Namespace, ResponseHandler> responseHandlers = new HashMap<>();
+
+        for (BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>> element : batch) {
+            NamespaceAndRequestParams requestParams = element.argument();
+            Namespace namespace = requestParams.namespace();
+            responseHandlers
+                    .computeIfAbsent(
+                            namespace,
+                            _unused ->
+                                    new ResponseHandler(requestParams.params().lockCleanupService()))
+                    .addPendingFuture(SettableResponse.of(requestParams.params().numTransactions(), element.result()));
+        }
+
+        return responseHandlers;
+    }
+
+    @VisibleForTesting
+    static Map<Namespace, RequestParams> getNamespaceWiseRequestParams(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        return batch.stream()
+                .map(BatchElement::argument)
+                .collect(Collectors.toMap(
+                        NamespaceAndRequestParams::namespace, NamespaceAndRequestParams::params, RequestParams::merge));
+    }
+
+    private static Map<Namespace, List<StartIdentifiedAtlasDbTransactionResponse>> getStartTransactionResponses(
+            Map<Namespace, RequestParams> originalRequestMap,
+            InternalMultiClientConjureTimelockService delegate,
+            UUID requestorId) {
+
+        Map<Namespace, ConjureStartTransactionsRequest> namespaceWiseRequests =
+                getNamespaceWiseRequests(originalRequestMap, requestorId);
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap = getResponseMap(delegate, namespaceWiseRequests);
+        return KeyedStream.stream(responseMap)
+                .mapEntries((namespace, response) -> {
+                    TransactionStarterHelper.updateCacheWithStartTransactionResponse(
```

Comment:
This is a side-effect inside a stream, which is not really good practice. I suggest you just rewrite this using a for loop

Reply:
Also the `getStartTransactionResponses` should indicate it's updating the cache